src/odbc_connection.cpp

```diff
@@ -2782,108 +2840,334 @@ fetch_and_store
-                  // a null terminator. Need to add an aditional byte.
-                  row[column_index].char_data = new SQLCHAR[row[column_index].size + 1]();
-                  memcpy
-                  (
-                    row[column_index].char_data,
-                    memory_start,
-                    row[column_index].size
-                  );
-                  break;
+                  // If not SQL_NO_TOTAL, we know how large to resize to buffer
+                  // to get the rest of the data.
+                  if (string_length_or_indicator != SQL_NO_TOTAL)
+                  {
+                    last_get_data_call = true;
+                  }
+
+                  // Handling the data that was returned depends heavily on the
+                  // target type of the binding data.
+                  switch(data->columns[column_index]->bind_type)
+                  {
+                    case SQL_C_BINARY:
+                      data_returned_length = buffer_size;
+                      buffer_size =
+                        string_length_or_indicator == SQL_NO_TOTAL ?
+                        buffer_size * 2 :
+                        row[column_index].size + string_length_or_indicator;
+                      row[column_index].char_data =
+                        (SQLCHAR *)
+                        realloc
+                        (
+                          row[column_index].char_data,
+                          buffer_size
+                        );
+                      row[column_index].size += data_returned_length;
+                      target_buffer =
+                        row[column_index].char_data + row[column_index].size;
+                    case SQL_C_WCHAR:
+                    {
+                      data_returned_length =
+                        strlen16
+                        (
+                          (const char16_t *) target_buffer
+                        ) * 2;
+                      buffer_size =
+                        string_length_or_indicator == SQL_NO_TOTAL ?
+                        buffer_size * 2 :
+                        row[column_index].size + string_length_or_indicator + 2;
+                      SQLWCHAR *temp_realloc =
+                        (SQLWCHAR *)
+                        realloc
+                        (
+                          row[column_index].wchar_data,
+                          buffer_size
+                        );
+                      if (temp_realloc == NULL)
+                      {
+                        free(row[column_index].wchar_data);
+                        *alloc_error = true;
+                        return return_code;
+                      } 
+                      row[column_index].wchar_data = temp_realloc;
+                      row[column_index].size += data_returned_length;
+                      target_buffer =
+                        row[column_index].wchar_data + row[column_index].size;
```

Comment:
Is size in terms of bytes of code units? If bytes, it needs to be divided by sizeof(SQLWCHAR) as adding a length to a pointer is equivalent to indexing.

Reply:
`size` is heavily correlated to `string_length_or_indicator` and `buffer_size`, both of which are in listed as dealing with "bytes" in the documentation. So yeah, need to divide here. Would love to avoid the "multiply by 2" after `strlen16` and the "divide by 2" here, but I think converting everything to bytes is going to be simplest.